### ISSUE TITLE
Provide configuration of exclusions via a list of exclusion rules that support per-line regular expressions, regular expressions against linter text, and by path

### ISSUE THREAD

**Issue by james-johnston-thumbtack** [PyCQA/flake8#1815]
### describe the request

flake8 allows you to disable certain rules in some files using [per-file-ignores](https://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-per-file-ignores).  This still has limited flexibility in that we can't disable certain rules for certain lines, e.g. matching a regular expression.

Compare with another similar linting tool for Golang that I regularly use, golangci-lint.  It provides an [issues configuration](https://golangci-lint.run/usage/configuration/#issues-configuration) that is very flexible for defining exclusions.  We can define various matching rules for exclusions (path, source code text/regular expression, linter output text/regular expression) in order to very flexibly exclude linter violations.

```yaml
  # Excluding configuration per-path, per-linter, per-text and per-source
  exclude-rules:
    # Exclude some linters from running on tests files.
    - path: _test\.go
      linters:
        - gocyclo
        - errcheck
        - dupl
        - gosec
    # Exclude known linters from partially hard-vendored code,
    # which is impossible to exclude via `nolint` comments.
    # `/` will be replaced by current OS file path separator to properly work on Windows.
    - path: internal/hmac/
      text: "weak cryptographic primitive"
      linters:
        - gosec
    # Exclude some `staticcheck` messages.
    - linters:
        - staticcheck
      text: "SA9003:"
    # Exclude `lll` issues for long lines with `go:generate`.
    - linters:
        - lll
      source: "^//go:generate "
```

Here is an example that I set up to ignore long lines based on a pattern I defined:
```yaml
  exclude-rules:
    # exclude long lines matching certain patterns:
    # - go:generate directives: there's no way to wrap these across multiple lines
    # - URLs: these really hurt readability to try to wrap
    # - github.com: a heuristic to exclude long import lines
    - linters:
        - revive
      text: "line-length-limit"  # match any linter error with this text in it (it's the name of the lint rule that was violated)
      # nice little regular expression to match the source code
      source: '^\s*//go:generate |https?://|github\.com'
```

I would like to accomplish something similar with flake8.  The pycodestyle plugin has a special case to ignore long URLs, but the rule is too rigid for my taste and I'd like to define a different pattern:
https://github.com/PyCQA/pycodestyle/blob/b22b672a1a53f1ef35c4f4689c674d66a24138ed/pycodestyle.py#L278-L285
Rather than have some special configurable thing with pycodestyle, I think it would be best if there was a general-purpose framework for ignoring issues similar to how other linter tools like golangci-lint can do.

**Comment by asottile** [comment]
no thanks, there's already designed ways to ignore things without all the additional complexity this would require

**Comment by james-johnston-thumbtack** [comment]
@asottile what is the recommended way to do this without requiring one to scatter `# noqa` lines everywhere as currently seems to be the case, or setting global/per-file ignores, while still minimizing the risk of hiding a true positive later on?

"there's already designated ways to ignore things" - did I miss something?  I only found per-file-ignores which is too broad since it applies to the whole file.  And "--ignore" just disables the rule everywhere.  I am generally not in favor of file-based exclusions (with exception of auto-generated code) because new & fixable issues could be introduced after the original person decided to ignore the whole file.  For example, person A says "I put 5 long URLs in this file, let me just set a per-file-ignore for E501", then person B comes along and dumps a bunch of random long lines in the file that could have been shortened because now they're allowed to.

Is there a more granular way to globally set an exclusion besides those flags I found, such as for the use case I outlined?
 Or is the permanent recommendation going forward to indeed always use a `# noqa` comment every time, even for cases that could be matched with a regex?

If the latter, I do feel that comment directives can also be problematic.  In my experience code reviewing, one common problem I've observed is forgetting to specify a code and making the directive to broad, e.g. `# noqa` will now ignore everything, not just the long line.  (hmmm, is there a lint rule to prevent broad noqa matches?)  But these sorts of problems can be avoided the more we can appropriately customize global linter behavior so that users don't need to worry about it as much in a project.

Regarding complexity: golangci-lint doesn't require cooperation from the individual linters for this.  All the lint rules run on all files and lines, like normal.  The rules get applied later on during the reporting phase, i.e. the lint violations get filtered out.  Thus they can just do it globally.  e.g. this is all the filtering logic they needed:
- https://github.com/golangci/golangci-lint/blob/b40a5443b3eddd748e1b1951222edf5b7a4f63b8/pkg/result/processors/exclude_rules.go
- filterIssues function: https://github.com/golangci/golangci-lint/blob/b673c5c1afce0db135afb52bc39374743ae43630/pkg/result/processors/issues.go#L9-L18
- rule matching logic: https://github.com/golangci/golangci-lint/blob/b40a5443b3eddd748e1b1951222edf5b7a4f63b8/pkg/result/processors/base_rule.go#L29-L49

A lot of that is boilerplate.... ultimately it really doesn't seem that bad?
